compiler/rustc_expand/src/expand.rs

```diff
@@ -716,7 +737,26 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
                 SyntaxExtensionKind::Attr(expander) => {
                     self.gate_proc_macro_input(&item);
                     self.gate_proc_macro_attr_item(span, &item);
-                    let tokens = item.into_tokens(&self.cx.sess.parse_sess);
+                    let tokens = match attr.style {
+                        AttrStyle::Outer => item.into_tokens(&self.cx.sess.parse_sess),
+                        // FIXME: Properly collect tokens for inner attributes
+                        AttrStyle::Inner => {
+                            let nt = match item {
+                                Annotatable::Item(i) => Nonterminal::NtItem(i),
+                                Annotatable::TraitItem(i) | Annotatable::ImplItem(i) => {
+                                    Nonterminal::NtItem(P(i.and_then(ast::AssocItem::into_item)))
+                                }
+                                Annotatable::ForeignItem(i) => {
+                                    Nonterminal::NtItem(P(i.and_then(ast::ForeignItem::into_item)))
+                                }
+                                Annotatable::Stmt(stmt) => Nonterminal::NtStmt(stmt.into_inner()),
+                                Annotatable::Expr(expr) => Nonterminal::NtExpr(expr),
+                                _ => unreachable!(),
+                            };
```

Comment:
This code can be deduplicated by factoring out `Annotatable::into_nonterminal` from `Annotatable::into_tokens`.